### PR TITLE
new collapsible_headings features

### LINKED ISSUE #616: [Feature Request] [Collapsible Headings] Shortcut to add heading above/below
A really slick UI feature that would save people a lot of annoyance would be a shortcut to add a heading at the current level above/below the current section. 

The most obvious use case is when you have two really long sections next to each other and want to add a section between. If you add a cell above the lower section title, the collapsible headings section (rightfully) expands the upper section. This then requires that you scroll to the bottom of the section to first add the desired section title, then usually back up again to collapse the accidentally expanded section. 

This seems silly, but in truly huge notebooks like the ones I use to organize my Ph.D. work, I end up temporarily reactivating Vimperator to navigate at that point, when collapsible-headings could in principle be enough to keep everything organized.

### PR DESCRIPTION
**Pull request by jcb91**
Fixes #616 and #609

### CODE REVIEW

**Comment by jcb91** [comment]
> Was going to submit a fix actually, or you can add it. You need a
> Jupyter.notebook.execute_cell() at line 595ish of main.js. That way, if you
> add a cell below, Collapsible Headings registers it as a title before you
> put the user into edit mode, preventing it from being "bury"'d inside of an
> existing, closed section.
> 
> Not sure if this makes sense?

Ah, does your new cell get hidden then? I thought I'd taken care of that by setting `select_reveals` false while inserting the new cell, but perhaps not?

File: nbextensions/usability/collapsible_headings/main.js [hunk 1]
```diff
@@ -533,6 +533,85 @@ define([
...
+		var cached_select_reveals = select_reveals;
+		select_reveals = false;
+		var new_cell = Jupyter.notebook.insert_cell_above('markdown', index);
+		// restore cached setting
+		select_reveals = cached_select_reveals;
+		var new_text = 'New heading';
+		new_cell.set_text(new_text);
+		new_cell.set_heading_level(level);
+		new_cell.code_mirror.setSelection({line:0, ch: level + 1}, {line:0, ch: level + 1 + new_text.length});
+		Jupyter.notebook.select(index, true);
```
Comment (brunobeltran) [inline nbextensions/usability/collapsible_headings/main.js]: adding a
`Jupyter.notebook.execute_cell()`
here after the `Jupyter.notebook.select(...)` would prevent the cell from being hidden when below, when done in combination with the current `select_reveals` code.

Comment (jcb91) [inline nbextensions/usability/collapsible_headings/main.js]: Ah, or, I should have put the select_reveals reset *after* the `Jupyter.notebook.select(index, true);` line - clue was in the name, really!

**Comment by brunobeltran** [comment]
Ah, left you a line comment then. The bug specifically occurs when inserting a new cell below a closed header cell. The previously existing header cell is reclosed by your `select_reveals` trick, but since you haven't executed the new cell yet, the plugin does not know that it is a title cell, just that it's a markdown cell, so it gets hidden at the bottom of the collapsed cell that you were trying to insert it beneath.

File: nbextensions/usability/collapsible_headings/main.js (hunk 1 above)
Comment (brunobeltran) [inline nbextensions/usability/collapsible_headings/main.js]: Yes, that works!

**Comment by jcb91** [comment]
Ok, I'm getting too tired to think about this clearly right now, so I think it's best to leave it for you to fix 😝 I guess the simplest way would be for you to submit a PR to the `feature/collapsible_headings` branch of my fork? I'm off to sleep, but will have a look at merging tomorrow

**Comment by jcb91** [comment]
as a side-effect, this also makes toc-cell links select the correct heading when collapsible headings is enabled 😝

**Comment by jcb91** [comment]
ok, I think I've successfully merged in #593 - let me know if/how well this works :)

**Comment by jcb91** [comment]
Ok, I think this works, and doesn't mangle toc2, so I'm going to merge it. Let me know if it breaks something!

**Pull request merged**